### PR TITLE
Attmpt to decouple this_code setup

### PR DESCRIPTION
**Pull request by unkcpz**
Hi @sphuber, this is an attempt to move the code script generate for `prepend_cmdline_params` ad `cmdline_params` as the parameters into `AbstractCode` class which is required by the containerized code implementation if the code type check needs to be removed from `presubmmit` of `calcjob.py`.

I add two helper function into `AbstractCode` to generate `prepend_cmdline_params` and `cmdline_params` with parameters. This is the pre-requisites of the containerized code implementation.

Please have a look and let me know if it is in the right direction. Thanks!

### CODE REVIEW

**Comment by unkcpz** [comment]
Hi @sphuber, I accept all the suggestions, should be fine now.

**Comment by unkcpz** [comment]
As we discussed offline, I did not implement to also remove the explicitly calling of `PortableCode` from `presubmit`, but for sure it needs to be removed and it is very confusing at `presubmit` that the `code` object from `inputs.all_nodes` and `this_code` from `calc_info` appears mixed in the function. I'll try to clarify it in another PR.

**Comment by unkcpz** [comment]
This PR is updated and tested in #5667, sufficient for that implementation.

File: aiida/engine/processes/calcjobs/calcjob.py [hunk 1]
```diff
@@ -734,11 +734,11 @@ def presubmit(self, folder: Folder) -> CalcInfo:
...
 
             if this_withmpi:
-                prepend_cmdline_params = mpi_args + extra_mpirun_params
+                prepend_cmdline_params = this_code.get_prepend_cmdline_params(mpi_args, extra_mpirun_params)
             else:
-                prepend_cmdline_params = []
+                prepend_cmdline_params = this_code.get_prepend_cmdline_params()
 
-            cmdline_params = [str(this_code.get_executable())] + (code_info.cmdline_params or [])
+            cmdline_params = this_code.get_executable_cmdline_params(cmdline_params=code_info.cmdline_params)
```
Comment (sphuber) [inline aiida/engine/processes/calcjobs/calcjob.py:741]:
```suggestion
            cmdline_params = this_code.get_executable_cmdline_params(code_info.cmdline_params)
```

File: aiida/orm/nodes/data/code/abstract.py [hunk 2]
```diff
@@ -70,12 +71,28 @@ def can_run_on_computer(self, computer: Computer) -> bool:
...
+    def get_executable(self) -> pathlib.Path:
         """Return the executable that the submission script should execute to run the code.
 
         :return: The executable to be called in the submission script.
         """
 
+    def get_executable_cmdline_params(self, cmdline_params: list[str] | None = None) -> list:
+        """Return the list of executable with its cmdline_params
+
+        :return: list of executable with its cmdline parameters."""
```
Comment (sphuber) [inline aiida/orm/nodes/data/code/abstract.py:83]:
```suggestion
        """Return the list of executable with its command line parameters.

        :param cmdline_params: List of command line parameters provided by the ``CalcJob`` plugin.
        :return: List of the executable followed by its command line parameters.
        """
```

File: aiida/orm/nodes/data/code/abstract.py [hunk 3]
```diff
@@ -70,12 +71,28 @@ def can_run_on_computer(self, computer: Computer) -> bool:
...
+        return [str(self.get_executable())] + (cmdline_params or [])
+
+    def get_prepend_cmdline_params( # pylint: disable=no-self-use
+        self,
+        mpi_args: list[str] | None = None,
+        extra_mpirun_params: list[str] | None = None
+    ) -> list[str]:
+        """Return the list of prepend cmdline params for mpi seeting
+
+        :return: list of prepend cmdline parameters."""
```
Comment (sphuber) [inline aiida/orm/nodes/data/code/abstract.py:93]:
```suggestion
        """Return List of command line parameters to be prepended to the executable in submission line.
        
        These command line parameters are typically parameters related to MPI invocations.

        :param mpi_args: List of MPI parameters provided by the ``Computer.get_mpirun_command`` method.
        :param extra_mpiruns_params: List of MPI parameters provided by the ``metadata.options.extra_mpirun_params`` input of the ``CalcJob``.
        :return: List of command line parameters to be prepended to the executable in submission line.
        """
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @unkcpz . If this works for the containerized PR, this is good to go. Just some suggestions for the docstrings. And I think there is still a call to `Code.get_executable` in the code that needs to cast to `str`:
https://github.com/aiidateam/aiida-core/blob/d666332e912d474ec729caf86acb2f6738225b39/aiida/orm/utils/builders/code.py#L114

**Comment by unkcpz** [comment]
@sphuber thanks a lot! All changed.

>  Just some suggestions for the docstrings. And I think there is still a call to Code.get_executable in the code that needs to cast to str.

True, and also change the type hint for the legacy code class and `MockCode` class for the test.

File: aiida/orm/nodes/data/code/abstract.py [hunk 4]
```diff
@@ -70,12 +71,38 @@ def can_run_on_computer(self, computer: Computer) -> bool:
         """
 
     @abc.abstractmethod
-    def get_executable(self) -> str:
-        """Return the executable that the submission script should execute to run the code.
+    def get_executable(self) -> pathlib.Path:
+        """
+        Return the executable that the submission script should execute to run the code.
```
Comment (sphuber) [inline aiida/orm/nodes/data/code/abstract.py:76]: PEP 257 docstring guidelines say to keep the first line on the same line as the opening quotes. Just a minor thing but good to keep the consistency when adding new code.
```suggestion
        """Return the executable that the submission script should execute to run the code.
```

File: aiida/orm/nodes/data/code/abstract.py [hunk 5]
```diff
@@ -70,12 +71,38 @@ def can_run_on_computer(self, computer: Computer) -> bool:
...
 
         :return: The executable to be called in the submission script.
         """
 
+    def get_executable_cmdline_params(self, cmdline_params: list[str] | None = None) -> list:
+        """
+        Return the list of executable with its command line parameters.
```
Comment (sphuber) [inline aiida/orm/nodes/data/code/abstract.py:83]:
```suggestion
    def get_executable_cmdline_params(self, cmdline_params: list[str] | None = None) -> list:
        """Return the list of executable with its command line parameters.
```

File: aiida/orm/nodes/data/code/abstract.py [hunk 6]
```diff
@@ -70,12 +71,38 @@ def can_run_on_computer(self, computer: Computer) -> bool:
...
+        """
+        return [str(self.get_executable())] + (cmdline_params or [])
+
+    def get_prepend_cmdline_params( # pylint: disable=no-self-use
+        self,
+        mpi_args: list[str] | None = None,
+        extra_mpirun_params: list[str] | None = None
+    ) -> list[str]:
+        """
+        Return List of command line parameters to be prepended to the executable in submission line.
```
Comment (sphuber) [inline aiida/orm/nodes/data/code/abstract.py:96]:
```suggestion
        """Return List of command line parameters to be prepended to the executable in submission line.
```

**Review by sphuber — CHANGES_REQUESTED** [review]
All good, if you can just update the branch (I have merged the other PR) and fix the minor comments on the docstrings

File: aiida/orm/nodes/data/code/abstract.py (hunk 4 above)
Comment (unkcpz) [reply aiida/orm/nodes/data/code/abstract.py]: I see, didn't know that. I thought you are too lazy to add a new line 😆

**Review by sphuber — APPROVED** [review]

**Pull request merged**